Adding practice problems to lists + a debugging callout (#28)
Ada-Developers-Academy/ada-precourse-v2: 01-intro-to-python/04-lists.md

```

## Practice Problems
<!--BEGIN CHALLENGE-->

### !challenge

* type: code-snippet
* language: python3.6
* id: 610f0431-b999-420f-8088-a40f5ddf9167
* title: Find Index of Item in List
* points: 1
* topics: python, lists

##### !question

Write a function `find_index_of_item` that takes two variables, an item and a list called list_of_items.  The function has the following behavior:
* If the list contains the item, the function returns the index of the item in the list.  
* If the list contains the item multiple times, the function returns the index of the first time the item appears in the list.  
* If the list does not contain the item, the function returns -1.

Example inputs and outputs:

input: ```find_index_of_item(3, [1, 4, 5, 6, 2, 3, 9])```  
output: ```5```

input: ```find_index_of_item("cat", ["dog", "cow", "goat", "pig"])```  
output: ```-1```

input: ```find_index_of_item("chocolate", [])```  
output: ```-1```

input: ```find_index_of_item(-93, [1, 30, -93, 99, -3, -93, 25, 16])```  
output: ```2```

##### !end-question

##### !placeholder

```python

def find_index_of_item(item, list_of_items):

>     return None

Author: I've updated all of the stub functions with pass.

```

##### !end-placeholder

##### !tests
```python

import unittest
from main import *

class TestFindIndexOfItemChallenge(unittest.TestCase):
    def test_find_success(self):
        self.assertEqual(find_index_of_item(3, [1, 4, 5, 6, 2, 3, 9]), 5)
    
    def test_find_first(self):
        self.assertEqual(find_index_of_item(-93, [1, 30, -93, 99, -3, -93, 25, 16]), 2)
    
    def test_find_failure(self):
        self.assertEqual(find_index_of_item("cat", ["dog", "cow", "goat", "pig"]), -1)
    
    def test_find_failure_empty_list(self):
        self.assertEqual(find_index_of_item("chocolate", []),-1)

```
##### !end-tests

<!--optional-->
##### !hint

A solution for this problem can be broken into the following steps:

1. Loop over list_of_items
1. Compare each element in list_of_items to item
1. If an element in list_of_items matches item, return the index of that element
1. If no matching element is found, return -1

##### !end-hint

<!--optional-->
##### !explanation

> Three examples of working implementations:

Author: Good call!  I've removed the enumerate solution & replaced it with this one!

    ```

### !callout-info

## Debugging

How do you know what your code is doing when it's generating a random word?  Use print() to print out the word during development.  Adding a print statement here will print the word to the terminal and make it easier to debug your code:
```python
snowman_word = r.word(
    word_min_length=SNOWMAN_MIN_WORD_LENGTH, 
    word_max_length=SNOWMAN_MAX_WORD_LENGTH)
print(snowman_word)
```

> Printing the value of variables in your code is an easy way to see what's going on inside your code.  Nobody writes perfect code!  There are always bugs, and learning how to debug is a core part of learning to be a programmer.  One strategy for debugging is to start by figuring out what your code is doing, and then work on making it do what you want.  Think of yourself as a detective, and instead of focusing on what you think your code should be doing, focus on figuring out exactly what it is doing.  Using all kinds of print statements is a simple way to start to look inside your code as it's running and see what's happening.

Reviewer: dang love this detective analogy

```suggestion
Printing the value of variables in your code is an easy way to see what's going on inside your code.  Nobody writes perfect code!  There are always bugs, and learning how to debug is a core part of learning to be a programmer.  One strategy for debugging is to start by figuring out what your code is doing, and then work on making it do what you want.  Think of yourself as a detective, and instead of focusing on what you think your code _should be_ doing, focus on figuring out exactly what it _is_ doing.  Using all kinds of print statements is a simple way to start looking inside your code as it's running and see what's happening.
```

### !end-callout

### Tracking User Input

> So far all we have done with our user input is check to see if it is in our word, but if we go back to the hypothetical game of snowman with a group, we would want to keep track of the letters that had been guessed.  We would also not accept guesses of the same letter that had been guessed before.  

Reviewer: ```suggestion
So far, all we have done with our user input is check to see if it is in our word. But if we go back to the hypothetical game of snowman with a group, we would want to keep track of the letters that have been guessed.  We would also not accept guesses of the same letter that have already been guessed.  
```
